### PR TITLE
Fix `TabControl` `DataContext` issues when switching tabs

### LINKED ISSUE #10846: TabControl: Null DataContext binding values being pushed when switching tabs
**Describe the bug**
Say we have a MainWindow with a TabControl containing 2 TabItems. Each TabItem's content is a view where the DataContext is bound to a viewModel that the MainWindow's viewModel contain. Everytime the tab is changed, an error gets logged to trace like the following: 
```
[Binding]Error in binding to 'Sandbox.FirstView'.'DataContext': 'Null value in expression '{empty}' at ''.'(FirstView #33726620)
```

**To Reproduce**
In the Sandbox project, prepare the following:
- Have the App.xaml.cs's framework initialise method be the following:
```cs
public override void OnFrameworkInitializationCompleted()
        {
            if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopLifetime)
            {
                desktopLifetime.MainWindow = new MainWindow()
                {
                    DataContext = new MainWindowViewModel()
                };
            }
            base.OnFrameworkInitializationCompleted();
            Trace.Listeners.Add(new TextWriterTraceListener(Console.Out));
            Trace.AutoFlush = true;
        }
```
MainWindow.xaml:
```cs
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        xmlns:s="clr-namespace:Sandbox"
        x:CompileBindings="False"
        x:DataType="s:MainWindowViewModel"
        RequestedThemeVariant="Dark"
        x:Class="Sandbox.MainWindow">
  <TabControl>
    <TabItem Header="tab 1">
      <s:FirstView DataContext="{Binding FirstVm}"/>
    </TabItem>
    <TabItem Header="tab 1">
      <s:SecondView DataContext="{Binding SecondVm}"/>
    </TabItem>
  </TabControl>
</Window>
```
MainWindowViewModel.cs:
```cs
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Sandbox;

public class MainWindowViewModel : INotifyPropertyChanged
{
    private FirstViewModel _firstVm;
    private SecondViewModel _secondVm;

    public FirstViewModel FirstVm
    {
        get => _firstVm;
        set
        {
            _firstVm = value;
            OnPropertyChanged();
        }
    }
    
    
    public SecondViewModel SecondVm
    {
        get => _secondVm;
        set
        {
            _secondVm = value;
            OnPropertyChanged();
        }
    }

    public MainWindowViewModel()
    {
        _firstVm = new FirstViewModel();
        _secondVm = new SecondViewModel();
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}
```
Finally, create the FirstView and SecondView as blank UserControl and their viewModel as blank classes. They don't need to contain anything, just be present.

Once you have everything setup, build the project, run it and then switch tabs. You should see one binding error everytime the current tab changes.

**Expected behavior**
To not see binding errors when switching tabs.

**Desktop (please complete the following information):**

- OS: Arch Linux `6.2.6-arch1-1 #1 SMP PREEMPT_DYNAMIC Mon, 13 Mar 2023 17:02:08 +0000 x86_64 GNU/Linux`
- Version Current master branch, but I also tested preview5 with the same issue

**Additional context**
Interestingly, I was in the idea this issue was related to https://github.com/AvaloniaUI/Avalonia/issues/9438 because I was experiencing this issue and the one I am reporting at the same time and they felt similar, but I have yet to investigate either deeply. It's possible to adapt the project such that both issues are reproduced at the same time though.

### LINKED ISSUE #18280: UserControl inside TabControl's TabItem is loosing its DataContext on Selected Tab is changed
### Describe the bug

When you have a TabControl with several TabItems and a UserControl as TabItem's Content, the UserControl's DataContext is set to null when the TabItem is changed.

In my case, i have a TabItem with a UserControl with DataGrid inside. The DataGrid's ItemSource is bound to a list (lets say a list of contacts) and the SelectedItem is bound to a property (lets say SelectedContact). When i now change the Tab, the DataContext of the UserControl is set to null and thus the SelectedItem of the DataGrid is propagating a null value to the SelectedContact property.

### To Reproduce

**ViewModels:**

AddressesViewModel

```
public class AddressesViewModel
{
	AddressViewModel _SelectedAddress;
	public AddressViewModel SelectedAddress
	{ 
		get => _SelectedAddress; 
		set => this.RaiseAndSetIfChanged(ref _SelectedAddress, value); 
	}

	ReadOnlyObservableCollection<AddressViewModel> _Addresses;
	public ReadOnlyObservableCollection<AddressViewModel> Addresses 
	{ 
		get => _Addresses; 
		internal set => this.RaiseAndSetIfChanged(ref _Addresses, value); 
	}
}
```

AddressViewModel 

```
public class AddressViewModel
{
	ObservableCollection<ContactViewModel> _Contacts;
	public ObservableCollection<ContactViewModel> Contacts
	{
		get => _Contacts;
		private set => this.RaiseAndSetIfChanged(ref _Contacts, value);
	}

	ContactViewModel _SelectedContact;
	public ContactViewModel SelectedContact
	{
		get => _SelectedContact;
		set => this.RaiseAndSetIfChanged(ref _SelectedContact, value);
	}
}
```

ContactViewModel 

```
public class ContactViewModel
{
	public string FirstName { get; set; }
	public string LastName { get; set; }
}
```

**View:**

AddressesView.xaml

```
<UserControl xmlns="https://github.com/avaloniaui"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	xmlns:sv="using:MyApp.Views.Addresses.SubViews"
	x:Class="MyApp.Views.Addresses.AddressesView">

	<DockPanel>
		<TextBlock Text="{Binding SelectedAddress.SelectedContact.FirstName}" DockPanel.Dock="Top" />

		<TabControl>
			<TabItem Header="Tab1" x:Name="tabContacts">
				<sv:AddressesContacts DataContext="{Binding SelectedAddress}" />
			</TabItem>

			<TabItem Header="Tab2" />
		</TabControl>
	</DockPanel>
</UserControl>
```

AddressesContacts.xaml

```
<UserControl xmlns="https://github.com/avaloniaui"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	x:Class="MyApp.Views.Addresses.SubViews.AddressesContacts">

	<DataGrid ItemsSource="{Binding Contacts}" SelectedItem="{Binding SelectedContact}" />
</UserControl>
```

When you switch from Tab1 to Tab2, and a contact is selected, SelectedAddress.SelectedContact is set to null.

### Expected behavior

The SelectedAddress.SelectedContact should stay as it is.

### Avalonia version

11.2.3

### OS

Windows

### Additional context

I asume, this is because of

`private void UpdateSelectedContent(Control? container = null)`

in TabControls.cs - line 196+

```
...

container ??= ContainerFromIndex(SelectedIndex);
if (container != null)
{
	if (SelectedContentTemplate != SelectContentTemplate(container.GetValue(ContentTemplateProperty)))
	{
		// If the value of SelectedContentTemplate is about to change, clear it first. This ensures
		// that the template is not reused as soon as SelectedContent changes in the statement below
		// this block, and also that controls generated from it are unloaded before SelectedContent
		// (which is typically their DataContext) changes.
		SelectedContentTemplate = null;
	}
}

...
```

A workaround is, to wrap the UserControl inside the TabItem in another Control - f.e.:

```
<TabItem Header="Tab1" x:Name="tabContacts">
	<ScrollViewer>
		<sv:AddressesContacts DataContext="{Binding SelectedAddress}" />
	</ScrollViewer>
</TabItem>
```

### LINKED ISSUE #20845: DataContext Bindings on TabItem children do not work.
### Describe the bug

When a `DataContext` binding is placed on the child of a `TabItem`, the `DataContext` is null.

NOTE: This is very similar to #18280 but does _not_ appear to be the same issue (though the same workaround works).

### To Reproduce

In Sandbox add the following:

```
    public class MainWindowViewModel
    {
        public Tab1ViewModel Tab1 { get; set; } = new Tab1ViewModel();
    }

    public class Tab1ViewModel
    {
        public string Name { get; set; } = "Tab 1 message here";
    }
```

```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        xmlns:local="using:Sandbox"
        x:Class="Sandbox.MainWindow"
        x:DataType="local:MainWindowViewModel">
  <TabControl>
    <TabItem Header="Tab 1">
      <local:Tab1View DataContext="{Binding Tab1}" />
    </TabItem>
  </TabControl>
</Window>
```

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:local="using:Sandbox"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="Sandbox.Tab1View"
             x:DataType="local:Tab1ViewModel">
  <TextBlock Text="{Binding Name}"/>
</UserControl>
```

(You can get the repro from https://github.com/AvaloniaUI/Avalonia/commit/ec893c0680d9e1f30e76c18672d12118d6e4d5ef).

Run the program, in debug mode [this `Debug.Assert`](https://github.com/AvaloniaUI/Avalonia/blob/45b0eb33b86f8298be902ba72483f28cc6a9c0ad/src/Avalonia.Controls/TabControl.cs?plain=1#L332C1-L332C94) is hit. In release mode the Tab1View's `TextBlock` displays nothing because its `DataContext` is null.

### Expected behavior

_No response_

### Avalonia version

master/12.0.0-preview2

### OS

_No response_

### Additional context

Seems to have been introduced by #20541

### PR DESCRIPTION
**Pull request by grokys**
## Background

The root issue of a [number](https://github.com/AvaloniaUI/Avalonia/issues/10958) [of](https://github.com/AvaloniaUI/Avalonia/issues/10242) `TabControl` [issues](https://github.com/AvaloniaUI/Avalonia/issues/10846) is that when you have a `TabControl`, the content area is distinct from the `TabItem`, meaning that we have:

```
TabControl
  |- ItemsPresenter #PART_ItemsPresenter
    |- TabItem
    |- TabItem
  |- Panel
    |- Content Area #PART_SelectedContentHost (ContentPresenter) 
    |- Content Area #PART_SelectedContentHost2 (ContentPresenter) (second one recently added for transitions) 
```

When you set a data context on the `TabItem`, you expect the content area to inherit that data context from the selected tab item but this doesn't happen automatically as there's no inheritance hierarchy between the `TabItem` and the content area[^1].

#20541 tried to fix this, but in doing so introduced #20845. This PR aims to fix it for good (famous last words)

## What does the pull request do?

- Add `ContentPresenter.SetContentWithDataContext` to atomically set `Content` and `DataContext`, preventing an intermediate state where the content inherits the wrong DataContext from higher up the tree. Yes, this method is a nasty hack and has to set a field to cause a side-effect elsewhere in the class. No, I'm not happy with it. Better solutions welcome!
- Update `TabControl.UpdateSelectedContent` to use the new method when content is a `Control`
- Fix the `DataContext` subscription to not apply the new container's `DataContext` to old content during page transitions
- Adds quite a few more unit tests which document the discovery of ever more edge cases while trying to fix this issue

Fixes #10846 
Fixes #18280
Fixes #20845

[^1]: Yes, `TabControl` is broken by design and I wish we didn't have it :(

### CODE REVIEW

**Comment by grokys** [comment]
#13230 looks to be related, we should take a look before merging (I only just found it).

File: src/Avalonia.Controls/TabControl.cs [hunk 1]
```diff
@@ -297,20 +296,20 @@ private void UpdateSelectedContent(Control? container = null)
...
-                            _contentPresenter2.IsVisible = true;
-                            if (contentElement is not null && contentElement.DataContext != contentDataContext)
-                                _contentPresenter2.DataContext = contentDataContext;
+                            _contentPresenter2!.ContentTemplate = template;
+                            _contentPresenter2!.IsVisible = true;
+
+                            if (content is Control)
+                                _contentPresenter2.SetContentWithDataContext(content, container.DataContext);
+                            else
+                                _contentPresenter2.Content = content;
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/TabControl.cs:312]: Fixed in 8a1a364fe: only use `SetContentWithDataContext` when `template is null`, and updated the comment in `UpdateChild`.

File: src/Avalonia.Controls/TabControl.cs [hunk 2]
```diff
@@ -320,18 +319,10 @@ private void UpdateSelectedContent(Control? container = null)
...
-                                    contentElement.DataContext != contentDataContext)
-                                {
-                                    Debug.Assert(!contentElement.IsSet(DataContextProperty));
-                                    ContentPart.DataContext = contentDataContext;
-                                }
+                                if (content is Control)
+                                    ContentPart.SetContentWithDataContext(content, container.DataContext);
+                                else
+                                    ContentPart.Content = content;
                             }
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/TabControl.cs:327]: Fixed in 8a1a364fe: only use `SetContentWithDataContext` when `template is null`, and updated the comment in `UpdateChild`.

File: src/Avalonia.Controls/TabControl.cs [hunk 3]
```diff
@@ -342,6 +333,21 @@ private void UpdateSelectedContent(Control? container = null)
...
+                        // the presenter that is showing this container's content.
+                        if (_contentPresenter2 is { IsVisible: true })
+                        {
+                            if (_contentPresenter2.Content is Control)
+                                _contentPresenter2.DataContext = dc;
+                        }
+                        else if (ContentPart?.Content is Control)
+                        {
+                            ContentPart.DataContext = dc;
+                        }
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/TabControl.cs:354]: Fixed in 8a1a364fe: only use `SetContentWithDataContext` when `template is null`, and updated the comment in `UpdateChild`.

File: src/Avalonia.Controls/Presenters/ContentPresenter.cs [hunk 4]
```diff
@@ -484,8 +518,17 @@ private void UpdateChild(object? content)
...
             // Set the DataContext if the data isn't a control.
-            if (contentTemplate is { } || !(content is Control))
+            if (overrideDataContext.IsSet)
+            {
+                DataContext = overrideDataContext.Value;
+            }
+            else if (contentTemplate is { } || !(content is Control))
             {
                 DataContext = content;
             }
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/Presenters/ContentPresenter.cs]: Fixed in 8a1a364fe: only use `SetContentWithDataContext` when `template is null`, and updated the comment in `UpdateChild`.

File: tests/Avalonia.Controls.UnitTests/TabControlTests.cs [hunk 5]
```diff
@@ -1046,6 +1046,438 @@ private IDisposable Start()
...
+            // Switch tab — triggers transition
+            target.SelectedIndex = 1;
+            root.LayoutManager.ExecuteLayoutPass();
+
+            // The old content (tab1View) should NOT have received Tab2's DataContext.
+            Assert.DoesNotContain(viewModel.Tab2, oldContentDataContexts);
+        }
+
+        [Fact]
+        public void ContentTemplate_With_Control_Content_Should_Set_DataContext_To_Content()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/TabControlTests.cs:1357]: Feels like a rather strange scenario, but I double-checked and these tests pass on `master` so Copilot was right to spot this. One more edge case without a test that I hadn't considered!

**Review by MrJul — APPROVED** [review]
LGTM!

**Comment by robloo** [comment]
> Yes, TabControl is broken by design and I wish we didn't have it

Well, TabControl is pretty important and shouldn't be impossible to get right. I think there were some architectural mistakes though. As with all controls it seems the complexity comes from just a few features. In this case:

 1. The optimization to only materialize one TabItem at a time
 2. The animations / transitions

We should just materialize each tab's content and have separate ContentPresenters for each tab. There is so much complication driven by sharing one ContentPresenter for what are logically separate parts of the UI. Doing this would completely eliminate the need for this PR and I think is the correct approach overall.

The next complication we added for questionable reasons is the duplicate ContentControl just for transitions -- we really should NOT have two. This itself feels quite hacky. IMO the animation done for the tabs needs to be completely re-thought. Maybe with connected animations or something one TabItem's content presenter can slide out while the next slides in.

**Comment by grokys** [comment]
> Well, TabControl is pretty important and shouldn't be impossible to get right.

That's where we disagree - IMO it _is_ impossible to get right because a `TabItem` control represents two disjoint controls - the header and the content. There is no other control where this is the case, and no sensibly architected control would do this.

**Comment by robloo** [comment]
> IMO it is impossible to get right because a TabItem control represents two disjoint controls - the header and the content

That's a pretty standard combination in all UI frameworks though. HeaderContentControl is the same abstraction. GroupBox is the same concept. Even a CheckBox is similar.

In this case there is a single source of truth -- the TabItems themselves. Then those are represented two ways 
 1) as the TabStrip for selection 
 2) as the visible content

If you treated both the TabStrip and the content-area as fundamentally the same data just different visualizations the issues go away. Both are just bound to the TabItems collections and have different data templates.

**Comment by Flithor** [comment]
I took a fall about this on 11.3.13, but currently not have a plan to upgrade to 12.0.
What is the workaround about this?

-----
Well, use `ReflectionBinding`. Even though it give a warning.

----
Add `x:DataType="UpperVMType"` on control can suppress warning.
